src/sqlancer/duckdb/test/DuckDBCODDTestOracle.java

```diff
@@ -0,0 +1,1065 @@
+import sqlancer.common.oracle.CODDTestBase;
+import sqlancer.common.oracle.TestOracle;
+import sqlancer.duckdb.DuckDBErrors;
+import sqlancer.duckdb.DuckDBProvider.DuckDBGlobalState;
+import sqlancer.duckdb.DuckDBSchema;
+import sqlancer.duckdb.DuckDBSchema.DuckDBColumn;
+import sqlancer.duckdb.DuckDBSchema.DuckDBCompositeDataType;
+import sqlancer.duckdb.DuckDBSchema.DuckDBTable;
+import sqlancer.duckdb.DuckDBSchema.DuckDBTables;
+import sqlancer.duckdb.DuckDBToStringVisitor;
+import sqlancer.duckdb.ast.DuckDBAlias;
+import sqlancer.duckdb.ast.DuckDBBinaryOperator;
+import sqlancer.duckdb.ast.DuckDBColumnReference;
+import sqlancer.duckdb.ast.DuckDBConstant;
+import sqlancer.duckdb.ast.DuckDBConstant.DuckDBTextConstant;
+import sqlancer.duckdb.ast.DuckDBExistsOperator;
+import sqlancer.duckdb.ast.DuckDBExpression;
+import sqlancer.duckdb.ast.DuckDBExpressionBag;
+import sqlancer.duckdb.ast.DuckDBFunction;
+import sqlancer.duckdb.ast.DuckDBInOperator;
+import sqlancer.duckdb.ast.DuckDBJoin;
+import sqlancer.duckdb.ast.DuckDBJoin.OuterType;
+import sqlancer.duckdb.ast.DuckDBOrderingTerm;
+import sqlancer.duckdb.ast.DuckDBResultMap;
+import sqlancer.duckdb.ast.DuckDBSelect;
+import sqlancer.duckdb.ast.DuckDBTableReference;
+import sqlancer.duckdb.ast.DuckDBTypeCast;
+import sqlancer.duckdb.ast.DuckDBTypeofNode;
+import sqlancer.duckdb.ast.DuckDBValues;
+import sqlancer.duckdb.ast.DuckDBWithClause;
+import sqlancer.duckdb.gen.DuckDBExpressionGenerator;
+import sqlancer.duckdb.gen.DuckDBExpressionGenerator.DuckDBAggregateFunction;
+import sqlancer.duckdb.gen.DuckDBExpressionGenerator.DuckDBBinaryArithmeticOperator;
+import sqlancer.duckdb.gen.DuckDBExpressionGenerator.DuckDBBinaryComparisonOperator;
+import sqlancer.duckdb.gen.DuckDBExpressionGenerator.DuckDBBinaryLogicalOperator;
+
+public class DuckDBCODDTestOracle extends CODDTestBase<DuckDBGlobalState> implements TestOracle<DuckDBGlobalState> {
+
+    private final DuckDBSchema s;
+    private DuckDBExpressionGenerator gen;
+    private Reproducer<DuckDBGlobalState> reproducer;
+
+    private static final String TEMP_TABLE_NAME = "temp_table";
+
+    private DuckDBExpression foldedExpr;
+    private DuckDBExpression constantResOfFoldedExpr;
+
+    private List<DuckDBTable> tablesFromOuterContext = new ArrayList<>();
+    private List<DuckDBJoin> joinsInExpr;
+
+    Map<String, List<DuckDBExpression>> auxiliaryQueryResult = new HashMap<>();
+    Map<String, List<DuckDBExpression>> selectResult = new HashMap<>();
+
+    Boolean useSubqueryAsFoldedExpr;
+    Boolean useCorrelatedSubqueryAsFoldedExpr;
+
+    public DuckDBCODDTestOracle(DuckDBGlobalState globalState) {
+        super(globalState);
+        this.s = globalState.getSchema();
+        DuckDBErrors.addExpressionErrors(errors);
+        DuckDBErrors.addInsertErrors(errors);
+        DuckDBErrors.addGroupByErrors(errors);
+    }
+
+    @Override
+    public void check() throws Exception {
+        reproducer = null;
```

Comment:
I feel like the logic in this class and method is difficult to understand. It would be ideal to refactor it, for example, by extracting methods and reducing code duplication between CODDTest implementations. However, I also understand that this might be risky, as we would no longer be certain that the implementation still works after refactoring.